google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/stub/CookiesHolderTest.java

```diff
@@ -0,0 +1,551 @@
+        new ServerInterceptor() {
+          @Override
+          public <ReqT, RespT> ServerCall.Listener<ReqT> interceptCall(
+              ServerCall<ReqT, RespT> serverCall,
+              Metadata metadata,
+              ServerCallHandler<ReqT, RespT> serverCallHandler) {
+            serverMetadata.add(metadata);
+            if (metadata.containsKey(ROUTING_COOKIE_1)) {
+              methods.add(serverCall.getMethodDescriptor().getBareMethodName());
+            }
+            return serverCallHandler.startCall(serverCall, metadata);
+          }
+        };
+
+    server = FakeServiceBuilder.create(fakeService).intercept(serverInterceptor).start();
+
+    BigtableDataSettings.Builder settings =
+        BigtableDataSettings.newBuilderForEmulator(server.getPort())
+            .setProjectId("fake-project")
+            .setInstanceId("fake-instance");
+
+    // Override CheckAndMutate and ReadModifyWrite retry settings here. These operations
+    // are currently not retryable but this could change in the future after we
+    // have routing cookie sends back meaningful information and changes how retry works.
+    // Routing cookie still needs to be respected and handled by the callables.
+    settings
+        .stubSettings()
+        .checkAndMutateRowSettings()
+        .setRetrySettings(
+            RetrySettings.newBuilder()
+                .setInitialRetryDelay(Duration.ofMillis(10))
+                .setMaxRetryDelay(Duration.ofMinutes(1))
+                .setMaxAttempts(2)
+                .build())
+        .setRetryableCodes(StatusCode.Code.UNAVAILABLE);
+
+    settings
+        .stubSettings()
+        .readModifyWriteRowSettings()
+        .setRetrySettings(
+            RetrySettings.newBuilder()
+                .setInitialRetryDelay(Duration.ofMillis(10))
+                .setMaxRetryDelay(Duration.ofMinutes(1))
+                .setMaxAttempts(2)
+                .build())
+        .setRetryableCodes(StatusCode.Code.UNAVAILABLE);
+
+    client = BigtableDataClient.create(settings.build());
+  }
+
+  @Test
+  public void testReadRows() {
+    client.readRows(Query.create("fake-table")).iterator().hasNext();
+
+    assertThat(fakeService.count.get()).isGreaterThan(1);
+    assertThat(serverMetadata.size()).isEqualTo(fakeService.count.get());
+    String bytes1 = serverMetadata.get(1).get(ROUTING_COOKIE_1);
+    String bytes2 = serverMetadata.get(1).get(ROUTING_COOKIE_2);
+    assertThat(bytes1).isNotNull();
+    assertThat(bytes1).isEqualTo("readRows");
+    assertThat(bytes2).isNotNull();
+    assertThat(bytes2).isEqualTo(testCookie);
+
+    // make sure bad key is not added
+    assertThat(serverMetadata.get(1).get(BAD_KEY)).isNull();
+
+    serverMetadata.clear();
+  }
+
+  @Test
+  public void testReadRow() {
+    client.readRow("fake-table", "key");
+
+    assertThat(fakeService.count.get()).isGreaterThan(1);
+    assertThat(serverMetadata.size()).isEqualTo(fakeService.count.get());
```

Comment:
nit. please use the appropriate checks for collections. ie `assertThat(serverMetadata).hasSize(fakeService.count.get());